File: crm_backend/routers/pipeline.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session

from ..database import get_db
from ..models import Pipeline, Stage, Lead, User, Booking
from ..auth import get_current_user

router = APIRouter()


@router.get("/")
def get_pipeline(
    current_user=Depends(get_current_user),
    db: Session = Depends(get_db),
):
    tenant_id = current_user.tenant_id
    pipeline = db.query(Pipeline).filter(
        Pipeline.tenant_id == tenant_id, Pipeline.is_default == True
    ).first()
    if not pipeline:
        raise HTTPException(status_code=404, detail="No pipeline found")

    stages = (
        db.query(Stage)
        .filter(Stage.pipeline_id == pipeline.id)
        .order_by(Stage.position)
        .all()
    )

    lead_rows = (
        db.query(
            Lead, User.name.label("assigned_name"),
            Booking.date.label("booking_date"), Booking.start_time.label("booking_start")
        )
        .outerjoin(User, Lead.assigned_to == User.id)
        .outerjoin(Booking, Booking.lead_id == Lead.id)
        .filter(Lead.tenant_id == tenant_id)
        .all()
    )

    lead_map: dict = {}
    for r in lead_rows:
        lead_map[r.Lead.stage_id] = lead_map.get(r.Lead.stage_id, [])
        lead_map[r.Lead.stage_id].append({
            "id": r.Lead.id,
            "first_name": r.Lead.first_name,
            "last_name": r.Lead.last_name,
            "company": r.Lead.company,
            "stage_id": r.Lead.stage_id,
            "status": r.Lead.status,
            "created_at": r.Lead.created_at,
            "assigned_to_name": r.assigned_name,
            "booking_date": str(r.booking_date) if r.booking_date else None,
            "booking_start": str(r.booking_start) if r.booking_start else None,
        })

    stages_with_leads = [
        {
            "id": s.id, "name": s.name, "color": s.color, "position": s.position,
            "leads": lead_map.get(s.id, []),
        }
        for s in stages
    ]

    return {
        "pipeline": {"id": pipeline.id, "name": pipeline.name},
        "stages": stages_with_leads,
    }

File: server/src/routes/pipeline.js
import { db } from "../config/database.js";
import { authenticate } from "../middleware/auth.js";

export default async function pipelineRoutes(fastify) {
  fastify.addHook("preHandler", authenticate);

  // GET /api/pipeline — all stages with leads
  fastify.get("/", async (request, reply) => {
    const tenantId = request.currentUser.tenant_id;
    const pipeline = await db("pipelines").where({ tenant_id: tenantId, is_default: true }).first();
    if (!pipeline) return reply.code(404).send({ error: "No pipeline found" });

    const stages = await db("stages")
      .where({ pipeline_id: pipeline.id })
      .orderBy("position", "asc");

    const leads = await db("leads")
      .where({ "leads.tenant_id": tenantId })
      .leftJoin("users", "leads.assigned_to", "users.id")
      .leftJoin("bookings", "bookings.lead_id", "leads.id")
      .select(
        "leads.id", "leads.first_name", "leads.last_name", "leads.company",
        "leads.stage_id", "leads.status", "leads.created_at",
        "users.name as assigned_to_name",
        "bookings.date as booking_date",
        "bookings.start_time as booking_start"
      );

    const stagesWithLeads = stages.map((stage) => ({
      ...stage,
      leads: leads.filter((l) => l.stage_id === stage.id),
    }));

    return reply.send({ pipeline, stages: stagesWithLeads });
  });
}
